File: android/app/src/main/java/com/practice/app/data/FirestoreRepository.kt
package com.practice.app.data

import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.SetOptions
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

data class PracticeSettings(
    val name: String = "Izadi Counselling",
    val welcomeSmsTemplate: String = defaultWelcomeSmsTemplate(),
    val currency: String = "AUD",
    val estimatedFeeDollars: Int = EstimatedSessionFeeDollars,
    val defaultFeeCents: Int = 15_000,
)

class FirestoreRepository(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance(),
) {
    private fun practiceRef(uid: String) = db.collection("practices").document(uid)
    private fun clientsRef(uid: String) = practiceRef(uid).collection("clients")
    private fun sessionsRef(uid: String) = practiceRef(uid).collection("sessions")

    fun currentUid(): String? = auth.currentUser?.uid

    suspend fun ensurePracticeDocument(uid: String, settings: PracticeSettings = PracticeSettings()) {
        val snap = practiceRef(uid).get().await()
        if (!snap.exists()) {
            practiceRef(uid).set(settings.toMap()).await()
        }
    }

    fun observeClients(uid: String): Flow<List<Client>> = callbackFlow {
        val registration: ListenerRegistration = clientsRef(uid)
            .orderBy("createdAtEpochMs", com.google.firebase.firestore.Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val clients = snapshot?.documents?.mapNotNull { it.toClientOrNull() }.orEmpty()
                trySend(clients)
            }
        awaitClose { registration.remove() }
    }

    fun observeSessions(uid: String): Flow<List<Session>> = callbackFlow {
        val registration: ListenerRegistration = sessionsRef(uid)
            .orderBy("startsAt")
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val sessions = snapshot?.documents?.mapNotNull { it.toSessionOrNull() }.orEmpty()
                trySend(sessions)
            }
        awaitClose { registration.remove() }
    }

    fun observeSettings(uid: String): Flow<PracticeSettings> = callbackFlow {
        val registration = practiceRef(uid).addSnapshotListener { snapshot, error ->
            if (error != null) {
                close(error)
                return@addSnapshotListener
            }
            trySend(snapshot?.data.toPracticeSettings())
        }
        awaitClose { registration.remove() }
    }

    suspend fun upsertClient(uid: String, client: Client) {
        clientsRef(uid).document(client.id).set(client.toMap(), SetOptions.merge()).await()
    }

    suspend fun deleteClient(uid: String, clientId: String) {
        val batch = db.batch()
        batch.delete(clientsRef(uid).document(clientId))
        val sessionDocs = sessionsRef(uid).whereEqualTo("clientId", clientId).get().await()
        sessionDocs.documents.forEach { batch.delete(it.reference) }
        batch.commit().await()
    }

    suspend fun upsertSession(uid: String, session: Session) {
        sessionsRef(uid).document(session.id).set(session.toMap(), SetOptions.merge()).await()
    }

    suspend fun updateWelcomeSmsTemplate(uid: String, template: String) {
        val saved = template.trim().ifEmpty { defaultWelcomeSmsTemplate() }
        practiceRef(uid).set(
            mapOf(
                "welcomeSmsTemplate" to saved,
                "updatedAt" to FieldValue.serverTimestamp(),
            ),
            SetOptions.merge(),
        ).await()
    }

    /** One-time import of on-device JSON into Firestore (merge by document id). */
    suspend fun importLocalData(uid: String, clients: List<Client>, sessions: List<Session>, template: String) {
        ensurePracticeDocument(
            uid,
            PracticeSettings(welcomeSmsTemplate = template.ifBlank { defaultWelcomeSmsTemplate() }),
        )
        val batchSize = 400
        clients.chunked(batchSize).forEach { chunk ->
            val batch = db.batch()
            chunk.forEach { client ->
                batch.set(clientsRef(uid).document(client.id), client.toMap(), SetOptions.merge())
            }
            batch.commit().await()
        }
        sessions.chunked(batchSize).forEach { chunk ->
            val batch = db.batch()
            chunk.forEach { session ->
                batch.set(sessionsRef(uid).document(session.id), session.toMap(), SetOptions.merge())
            }
            batch.commit().await()
        }
        updateWelcomeSmsTemplate(uid, template)
    }
}

private fun Client.toMap(): Map<String, Any?> = mapOf(
    "firstName" to firstName,
    "surname" to surname,
    "dateOfBirth" to dateOfBirth,
    "mobile" to mobile,
    "emergencyContactName" to emergencyContactName,
    "emergencyContactNumber" to emergencyContactNumber,
    "relationship" to relationship?.name,
    "isActive" to isActive,
    "createdAtEpochMs" to createdAtEpochMs,
)

private fun com.google.firebase.firestore.DocumentSnapshot.toClientOrNull(): Client? {
    val data = data ?: return null
    val firstName = data["firstName"] as? String ?: return null
    val surname = data["surname"] as? String ?: return null
    val relationshipName = data["relationship"] as? String
    return Client(
        id = id,
        firstName = firstName,
        surname = surname,
        dateOfBirth = data["dateOfBirth"] as? String ?: "",
        mobile = data["mobile"] as? String ?: "",
        emergencyContactName = data["emergencyContactName"] as? String ?: "",
        emergencyContactNumber = data["emergencyContactNumber"] as? String ?: "",
        relationship = relationshipName?.let { name ->
            EmergencyRelationship.entries.find { it.name == name }
        },
        isActive = data["isActive"] as? Boolean ?: true,
        createdAtEpochMs = (data["createdAtEpochMs"] as? Number)?.toLong() ?: 0L,
    )
}

private fun Session.toMap(): Map<String, Any?> {
    val zone = ZoneId.systemDefault()
    val starts = startsAt.atZone(zone).toInstant()
    val ends = endsAt.atZone(zone).toInstant()
    return mapOf(
        "clientId" to clientId,
        "clientName" to clientName,
        "date" to date.toString(),
        "time" to time.toString().take(5),
        "startsAt" to Timestamp(starts.epochSecond, starts.nano),
        "endsAt" to Timestamp(ends.epochSecond, ends.nano),
        "durationMinutes" to durationMinutes,
        "notes" to notes,
        "status" to status.name,
        "paymentStatus" to paymentStatus.name,
        "paymentAmountCents" to paymentAmountCents,
        "paymentMethod" to paymentMethod?.name,
        "updatedAt" to FieldValue.serverTimestamp(),
    )
}

private fun com.google.firebase.firestore.DocumentSnapshot.toSessionOrNull(): Session? {
    val data = data ?: return null
    val clientId = data["clientId"] as? String ?: return null
    val clientName = data["clientName"] as? String ?: return null
    val date = runCatching { LocalDate.parse(data["date"] as? String ?: return null) }.getOrNull()
        ?: return null
    val time = runCatching { LocalTime.parse(data["time"] as? String ?: return null) }.getOrNull()
        ?: return null
    val duration = (data["durationMinutes"] as? Number)?.toInt() ?: return null
    val statusRaw = data["status"] as? String ?: SessionStatus.Scheduled.name
    val status = when (statusRaw) {
        "Postponed" -> SessionStatus.Scheduled
        else -> SessionStatus.entries.find { it.name == statusRaw } ?: SessionStatus.Scheduled
    }
    val paymentRaw = data["paymentStatus"] as? String
    val paymentStatus = when (paymentRaw) {
        "NotPaid" -> PaymentStatus.Paid
        else -> paymentRaw?.let { name -> PaymentStatus.entries.find { it.name == name } }
            ?: if (status == SessionStatus.Completed) PaymentStatus.Paid else PaymentStatus.Pending
    }
    val method = (data["paymentMethod"] as? String)?.let { name ->
        PaymentMethod.entries.find { it.name == name }
    }
    return Session(
        id = id,
        clientId = clientId,
        clientName = clientName,
        date = date,
        time = time,
        durationMinutes = duration,
        notes = data["notes"] as? String ?: "",
        status = status,
        paymentStatus = paymentStatus,
        paymentAmountCents = (data["paymentAmountCents"] as? Number)?.toInt(),
        paymentMethod = method,
    )
}

private fun PracticeSettings.toMap(): Map<String, Any?> = mapOf(
    "name" to name,
    "welcomeSmsTemplate" to welcomeSmsTemplate,
    "currency" to currency,
    "estimatedFeeDollars" to estimatedFeeDollars,
    "defaultFeeCents" to defaultFeeCents,
    "updatedAt" to FieldValue.serverTimestamp(),
)

private fun Map<String, Any>?.toPracticeSettings(): PracticeSettings {
    if (this == null) return PracticeSettings()
    return PracticeSettings(
        name = this["name"] as? String ?: "Izadi Counselling",
        welcomeSmsTemplate = this["welcomeSmsTemplate"] as? String ?: defaultWelcomeSmsTemplate(),
        currency = this["currency"] as? String ?: "AUD",
        estimatedFeeDollars = (this["estimatedFeeDollars"] as? Number)?.toInt()
            ?: EstimatedSessionFeeDollars,
        defaultFeeCents = (this["defaultFeeCents"] as? Number)?.toInt() ?: 15_000,
    )
}

File: ios/Izadi/Izadi/Services/FirestoreRepository.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

enum FirestorePaths {
    static func practice(_ uid: String) -> DocumentReference {
        Firestore.firestore().collection("practices").document(uid)
    }

    static func clients(_ uid: String) -> CollectionReference {
        practice(uid).collection("clients")
    }

    static func sessions(_ uid: String) -> CollectionReference {
        practice(uid).collection("sessions")
    }
}

struct PracticeSettings: Equatable {
    var name: String
    var welcomeSmsTemplate: String
    var courtesySmsTemplate: String
    var currency: String
    var estimatedFeeDollars: Int
    var defaultFeeCents: Int

    static let `default` = PracticeSettings(
        name: "Izadi Counselling",
        welcomeSmsTemplate: WelcomeSms.defaultTemplate,
        courtesySmsTemplate: CourtesySms.defaultTemplate,
        currency: "AUD",
        estimatedFeeDollars: 130,
        defaultFeeCents: 15_000
    )
}

enum FirestoreMapping {
    static func client(from data: [String: Any], id: String) -> Client? {
        guard
            let firstName = data["firstName"] as? String,
            let surname = data["surname"] as? String
        else { return nil }

        let relationshipRaw = data["relationship"] as? String
        let relationshipCustomRaw = (data["relationshipCustom"] as? String ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let (relationship, relationshipCustom) = EmergencyRelationship.parseStored(
            relationship: relationshipRaw,
            relationshipCustom: relationshipCustomRaw
        )

        return Client(
            id: id,
            firstName: firstName,
            surname: surname,
            dateOfBirth: data["dateOfBirth"] as? String ?? "",
            mobile: data["mobile"] as? String ?? "",
            emergencyContactName: data["emergencyContactName"] as? String ?? "",
            emergencyContactNumber: data["emergencyContactNumber"] as? String ?? "",
            relationship: relationship,
            relationshipCustom: relationshipCustom,
            isActive: data["isActive"] as? Bool ?? true,
            createdAtEpochMs: (data["createdAtEpochMs"] as? NSNumber)?.int64Value
                ?? Int64(Date().timeIntervalSince1970 * 1000)
        )
    }

    static func clientData(_ client: Client) -> [String: Any] {
        [
            "firstName": client.firstName,
            "surname": client.surname,
            "dateOfBirth": client.dateOfBirth,
            "mobile": client.mobile,
            "emergencyContactName": client.emergencyContactName,
            "emergencyContactNumber": client.emergencyContactNumber,
            "relationship": client.relationship?.rawValue as Any,
            "relationshipCustom": client.relationship == .other
                ? client.relationshipCustom.trimmingCharacters(in: .whitespacesAndNewlines)
                : "",
            "isActive": client.isActive,
            "createdAtEpochMs": client.createdAtEpochMs,
        ]
    }

    static func session(from data: [String: Any], id: String) -> Session? {
        guard
            let clientId = data["clientId"] as? String,
            let clientName = data["clientName"] as? String,
            let dateString = data["date"] as? String,
            let timeString = data["time"] as? String,
            let duration = data["durationMinutes"] as? Int
        else { return nil }

        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: dateString),
              let timeMinutes = TimeHelpers.parseTimeString(timeString)
        else { return nil }

        let statusRaw = data["status"] as? String ?? SessionStatus.scheduled.rawValue
        let status: SessionStatus = {
            if statusRaw == "Postponed" { return .scheduled }
            return SessionStatus(rawValue: statusRaw) ?? .scheduled
        }()

        let paymentRaw = data["paymentStatus"] as? String
        let paymentStatus: PaymentStatus = {
            if paymentRaw == "NotPaid" { return .paid }
            if let paymentRaw, let parsed = PaymentStatus(rawValue: paymentRaw) {
                return parsed
            }
            return status == .completed ? .paid : .pending
        }()

        let methodRaw = data["paymentMethod"] as? String
        let method = methodRaw.flatMap(PaymentMethod.init(rawValue:))
        let amount = data["paymentAmountCents"] as? Int

        return Session(
            id: id,
            clientId: clientId,
            clientName: clientName,
            date: date,
            timeMinutes: timeMinutes,
            durationMinutes: duration,
            notes: data["notes"] as? String ?? "",
            status: status,
            paymentStatus: paymentStatus,
            paymentAmountCents: amount,
            paymentMethod: method,
            courtesySmsCompleted: data["courtesySmsCompleted"] as? Bool ?? false
        )
    }

    static func sessionData(_ session: Session) -> [String: Any] {
        [
            "clientId": session.clientId,
            "clientName": session.clientName,
            "date": session.dateString,
            "time": session.timeString,
            "startsAt": Timestamp(date: session.startsAt),
            "endsAt": Timestamp(date: session.endsAt),
            "durationMinutes": session.durationMinutes,
            "notes": session.notes,
            "status": session.status.rawValue,
            "paymentStatus": session.paymentStatus.rawValue,
            "paymentAmountCents": session.paymentAmountCents as Any,
            "paymentMethod": session.paymentMethod?.rawValue as Any,
            "courtesySmsCompleted": session.courtesySmsCompleted,
            "updatedAt": FieldValue.serverTimestamp(),
        ]
    }

    static func settings(from data: [String: Any]?) -> PracticeSettings {
        guard let data else { return .default }
        return PracticeSettings(
            name: data["name"] as? String ?? PracticeSettings.default.name,
            welcomeSmsTemplate: data["welcomeSmsTemplate"] as? String ?? WelcomeSms.defaultTemplate,
            courtesySmsTemplate: data["courtesySmsTemplate"] as? String ?? CourtesySms.defaultTemplate,
            currency: data["currency"] as? String ?? "AUD",
            estimatedFeeDollars: data["estimatedFeeDollars"] as? Int ?? 130,
            defaultFeeCents: data["defaultFeeCents"] as? Int ?? 15_000
        )
    }

    static func settingsData(_ settings: PracticeSettings) -> [String: Any] {
        [
            "name": settings.name,
            "welcomeSmsTemplate": settings.welcomeSmsTemplate,
            "courtesySmsTemplate": settings.courtesySmsTemplate,
            "currency": settings.currency,
            "estimatedFeeDollars": settings.estimatedFeeDollars,
            "defaultFeeCents": settings.defaultFeeCents,
            "updatedAt": FieldValue.serverTimestamp(),
        ]
    }
}

final class FirestoreRepository {
    private var clientListener: ListenerRegistration?
    private var sessionListener: ListenerRegistration?
    private var settingsListener: ListenerRegistration?

    func ensurePracticeDocument(uid: String, settings: PracticeSettings = .default) async throws {
        let ref = FirestorePaths.practice(uid)
        let snapshot = try await ref.getDocument()
        if !snapshot.exists {
            try await ref.setData(FirestoreMapping.settingsData(settings))
        }
    }

    func listen(
        uid: String,
        onClients: @escaping ([Client]) -> Void,
        onSessions: @escaping ([Session]) -> Void,
        onSettings: @escaping (PracticeSettings) -> Void,
        onError: @escaping (Error) -> Void
    ) {
        stopListening()

        settingsListener = FirestorePaths.practice(uid).addSnapshotListener { snapshot, error in
            if let error {
                onError(error)
                return
            }
            onSettings(FirestoreMapping.settings(from: snapshot?.data()))
        }

        clientListener = FirestorePaths.clients(uid)
            .order(by: "createdAtEpochMs", descending: true)
            .addSnapshotListener { snapshot, error in
                if let error {
                    onError(error)
                    return
                }
                let clients = snapshot?.documents.compactMap {
                    FirestoreMapping.client(from: $0.data(), id: $0.documentID)
                } ?? []
                onClients(clients)
            }

        sessionListener = FirestorePaths.sessions(uid)
            .order(by: "startsAt", descending: false)
            .addSnapshotListener { snapshot, error in
                if let error {
                    onError(error)
                    return
                }
                let sessions = snapshot?.documents.compactMap {
                    FirestoreMapping.session(from: $0.data(), id: $0.documentID)
                } ?? []
                onSessions(sessions)
            }
    }

    func stopListening() {
        clientListener?.remove()
        sessionListener?.remove()
        settingsListener?.remove()
        clientListener = nil
        sessionListener = nil
        settingsListener = nil
    }

    func upsertClient(uid: String, client: Client) async throws {
        try await FirestorePaths.clients(uid)
            .document(client.id)
            .setData(FirestoreMapping.clientData(client), merge: true)
    }

    func deleteClient(uid: String, clientId: String) async throws {
        let db = Firestore.firestore()
        let batch = db.batch()
        batch.deleteDocument(FirestorePaths.clients(uid).document(clientId))

        let sessionDocs = try await FirestorePaths.sessions(uid)
            .whereField("clientId", isEqualTo: clientId)
            .getDocuments()
        for doc in sessionDocs.documents {
            batch.deleteDocument(doc.reference)
        }
        try await batch.commit()
    }

    func upsertSession(uid: String, session: Session) async throws {
        try await FirestorePaths.sessions(uid)
            .document(session.id)
            .setData(FirestoreMapping.sessionData(session), merge: true)
    }

    func deleteSession(uid: String, sessionId: String) async throws {
        try await FirestorePaths.sessions(uid).document(sessionId).delete()
    }

    func updateWelcomeSmsTemplate(uid: String, template: String) async throws {
        let saved = template.trimmingCharacters(in: .whitespacesAndNewlines)
        let value = saved.isEmpty ? WelcomeSms.defaultTemplate : saved
        try await FirestorePaths.practice(uid).setData(
            [
                "welcomeSmsTemplate": value,
                "updatedAt": FieldValue.serverTimestamp(),
            ],
            merge: true
        )
    }

    func updateCourtesySmsTemplate(uid: String, template: String) async throws {
        let saved = template.trimmingCharacters(in: .whitespacesAndNewlines)
        let value = saved.isEmpty ? CourtesySms.defaultTemplate : saved
        try await FirestorePaths.practice(uid).setData(
            [
                "courtesySmsTemplate": value,
                "updatedAt": FieldValue.serverTimestamp(),
            ],
            merge: true
        )
    }
}
